diff --git a/third_party/proton/Dialect/include/Dialect/Proton/IR/ProtonDialect.td b/third_party/proton/Dialect/include/Dialect/Proton/IR/ProtonDialect.td
index 8d35a5ec1..e476cad96 100644
--- a/third_party/proton/Dialect/include/Dialect/Proton/IR/ProtonDialect.td
+++ b/third_party/proton/Dialect/include/Dialect/Proton/IR/ProtonDialect.td
@@ -14,7 +14,6 @@ def Proton_Dialect : Dialect {

   let dependentDialects = [];

-  let usePropertiesForAttributes = 1;
 }

 #endif
diff --git a/third_party/proton/Dialect/include/Dialect/ProtonGPU/IR/ProtonGPUDialect.td b/third_party/proton/Dialect/include/Dialect/ProtonGPU/IR/ProtonGPUDialect.td
index 9fe146f5e..c63baa385 100644
--- a/third_party/proton/Dialect/include/Dialect/ProtonGPU/IR/ProtonGPUDialect.td
+++ b/third_party/proton/Dialect/include/Dialect/ProtonGPU/IR/ProtonGPUDialect.td
@@ -22,7 +22,6 @@ def ProtonGPU_Dialect : Dialect {

   let useDefaultTypePrinterParser = 1;
   let useDefaultAttributePrinterParser = 1;
-  let usePropertiesForAttributes = 1;
 }

 #endif  // PROTONGPU_DIALECT
diff --git a/third_party/tle/CMakeLists.txt b/third_party/tle/CMakeLists.txt
index 1034b0fef..6d419e1ef 100644
--- a/third_party/tle/CMakeLists.txt
+++ b/third_party/tle/CMakeLists.txt
@@ -23,4 +23,6 @@ if(TRITON_BUILD_PYTHON_MODULE)
   add_dependencies(TritonTLE TleTableGen TritonTLETransformsIncGen)
 endif()

-add_subdirectory(test)
+if(TRITON_BUILD_UT)
+  add_subdirectory(test)
+endif()
diff --git a/third_party/tle/dialect/include/IR/TleDialect.td b/third_party/tle/dialect/include/IR/TleDialect.td
index bb506128e..0fd8d841a 100644
--- a/third_party/tle/dialect/include/IR/TleDialect.td
+++ b/third_party/tle/dialect/include/IR/TleDialect.td
@@ -38,7 +38,6 @@ def Tle_Dialect : Dialect {
     "triton::TritonDialect",
     "triton::gpu::TritonGPUDialect",
   ];
-  let usePropertiesForAttributes = 1;
 }

 #endif
